massa-ledger-exports/src/key.rs

```diff
@@ -1,105 +1,191 @@
 use massa_models::{
-    address::{Address, AddressDeserializer, ADDRESS_SIZE_BYTES},
+    address::{Address, AddressDeserializer, AddressSerializer},
     serialization::{VecU8Deserializer, VecU8Serializer},
 };
-use massa_serialization::{DeserializeError, Deserializer, SerializeError, Serializer};
+use massa_serialization::{
+    Deserializer, SerializeError, Serializer, U64VarIntDeserializer, U64VarIntSerializer,
+};
 use nom::error::{ContextError, ParseError};
+use num_enum::{IntoPrimitive, TryFromPrimitive};
 use std::ops::Bound::Included;
 
 pub const BALANCE_IDENT: u8 = 0u8;
 pub const BYTECODE_IDENT: u8 = 1u8;
 pub const DATASTORE_IDENT: u8 = 2u8;
+pub const KEY_VERSION: u64 = 0;
+
+#[derive(PartialEq, Eq, Clone, IntoPrimitive, TryFromPrimitive, Debug)]
+#[repr(u8)]
+enum KeyTypeId {
+    Balance = 0,
+    Bytecode = 1,
+    Datastore = 2,
+}
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum KeyType {
+    BALANCE,
+    BYTECODE,
+    DATASTORE(Vec<u8>),
+}
+
+#[derive(Default, Clone)]
+pub struct KeyTypeSerializer {
+    vec_u8_serializer: VecU8Serializer,
+    // Whether is deserialized with VecU8Deserializer or not.
+    // If true, we use the VecU8Serializer to serialize the key which will add the length at the beginning.
+    // If false, we just serialize the key as is.
+    // This allows us to store the datastore key length at the beginning of the key or not.
+    // The datastore key length is useful when transfering multiple keys, like in packets,
+    // but isn't when storing a datastore key in the ledger.
+    with_datastore_key_length: bool,
```

Comment:
Premature optimization: given the added complexity and risk of error in the future if we forget to set this value properly, I would vote for the ledger size penalty over code fragmentation

Reply:
this needs a followup issue

Reply:
#3708

Reply:
thx